File: day_16_sampling/day_16_sampling.js
/*
 * Sampling: Population, Samples and Sampling Techniques
 *
 * This self-contained JavaScript program demonstrates:
 * - Population and sample
 * - Parameters and statistics
 * - Simple random sampling
 * - Systematic sampling
 * - Stratified sampling
 * - Cluster and multistage sampling
 * - Convenience and quota sampling
 * - Sample-size estimation
 * - Sampling error
 * - Confidence intervals
 * - Weighted estimation
 * - Repeated sampling and sampling distributions
 * - Validation and edge cases
 *
 * Compatible with modern Node.js.
 */

"use strict";

// ---------------------------------------------------------------------------
// 1. REPRODUCIBLE RANDOM NUMBER GENERATOR
// ---------------------------------------------------------------------------

// JavaScript's Math.random() cannot be seeded directly.
// A small deterministic generator makes demonstrations reproducible.
class SeededRandom {
    constructor(seed = 42) {
        this.state = seed >>> 0;
    }

    next() {
        this.state = (1664525 * this.state + 1013904223) >>> 0;
        return this.state / 4294967296;
    }

    integer(min, max) {
        return Math.floor(this.next() * (max - min + 1)) + min;
    }

    choice(array) {
        return array[this.integer(0, array.length - 1)];
    }

    shuffle(array) {
        const copy = [...array];

        for (let i = copy.length - 1; i > 0; i--) {
            const j = this.integer(0, i);
            [copy[i], copy[j]] = [copy[j], copy[i]];
        }

        return copy;
    }

    sample(array, size) {
        if (size < 0 || size > array.length) {
            throw new RangeError("Invalid sample size.");
        }

        return this.shuffle(array).slice(0, size);
    }
}

// ---------------------------------------------------------------------------
// 2. POPULATION MODEL
// ---------------------------------------------------------------------------

function buildPopulation(size = 3000, seed = 42) {
    if (!Number.isInteger(size) || size <= 0) {
        throw new RangeError("Population size must be positive.");
    }

    const random = new SeededRandom(seed);
    const regions = ["North", "South", "East", "West"];
    const employmentTypes = [
        "Student",
        "Employed",
        "Self-employed",
        "Unemployed"
    ];
    const population = [];

    for (let id = 1; id <= size; id++) {
        const region = weightedChoice(
            random,
            regions,
            [0.35, 0.25, 0.20, 0.20]
        );

        const employment = weightedChoice(
            random,
            employmentTypes,
            [0.15, 0.50, 0.25, 0.10]
        );

        const age = random.integer(18, 70);

        const incomeBase = {
            Student: 12000,
            Employed: 50000,
            "Self-employed": 60000,
            Unemployed: 10000
        }[employment];

        // Uniform noise is sufficient for this educational simulation.
        const income = Math.max(
            0,
            incomeBase + (random.next() - 0.5) * incomeBase
        );

        const satisfaction = clamp(
            6 + income / 200000 + (random.next() - 0.5) * 3,
            1,
            10
        );

        const cluster = `District-${String(((id - 1) % 15) + 1).padStart(2, "0")}`;

        population.push({
            id,
            age,
            region,
            employment,
            income,
            satisfaction,
            cluster
        });
    }

    return population;
}

function weightedChoice(random, values, weights) {
    const total = weights.reduce((sum, weight) => sum + weight, 0);
    let target = random.next() * total;

    for (let i = 0; i < values.length; i++) {
        target -= weights[i];

        if (target <= 0) {
            return values[i];
        }
    }

    return values[values.length - 1];
}

function clamp(value, minimum, maximum) {
    return Math.min(maximum, Math.max(minimum, value));
}

// ---------------------------------------------------------------------------
// 3. BASIC STATISTICS
// ---------------------------------------------------------------------------

function arithmeticMean(values) {
    if (values.length === 0) {
        throw new Error("Cannot calculate a mean of an empty array.");
    }

    return values.reduce((sum, value) => sum + value, 0) / values.length;
}

function sampleStandardDeviation(values) {
    if (values.length < 2) {
        return 0;
    }

    const average = arithmeticMean(values);
    const squaredDifferences = values.map(
        value => (value - average) ** 2
    );

    return Math.sqrt(
        squaredDifferences.reduce((sum, value) => sum + value, 0)
        / (values.length - 1)
    );
}

function median(values) {
    if (values.length === 0) {
        throw new Error("Cannot calculate a median of an empty array.");
    }

    const sorted = [...values].sort((a, b) => a - b);
    const middle = Math.floor(sorted.length / 2);

    return sorted.length % 2 === 0
        ? (sorted[middle - 1] + sorted[middle]) / 2
        : sorted[middle];
}

function validateSampleSize(populationSize, sampleSize) {
    if (!Number.isInteger(sampleSize) || sampleSize <= 0) {
        throw new RangeError("Sample size must be a positive integer.");
    }

    if (sampleSize > populationSize) {
        throw new RangeError(
            "Sample size cannot exceed population size."
        );
    }
}

// ---------------------------------------------------------------------------
// 4. SIMPLE RANDOM SAMPLING
// ---------------------------------------------------------------------------

function simpleRandomSample(population, sampleSize, seed = 42) {
    validateSampleSize(population.length, sampleSize);

    const random = new SeededRandom(seed);
    return random.sample(population, sampleSize);
}

// ---------------------------------------------------------------------------
// 5. SYSTEMATIC SAMPLING
// ---------------------------------------------------------------------------

function systematicSample(population, sampleSize, seed = 42) {
    validateSampleSize(population.length, sampleSize);

    const random = new SeededRandom(seed);
    const interval = population.length / sampleSize;
    const start = random.next() * interval;

    const result = [];

    for (let i = 0; i < sampleSize; i++) {
        const index = Math.min(
            population.length - 1,
            Math.floor(start + i * interval)
        );

        result.push(population[index]);
    }

    return result;
}

// ---------------------------------------------------------------------------
// 6. STRATIFIED SAMPLING
// ---------------------------------------------------------------------------

function stratifiedSample(
    population,
    sampleSize,
    attributeFunction,
    seed = 42
) {
    validateSampleSize(population.length, sampleSize);

    const random = new SeededRandom(seed);
    const strata = new Map();

    for (const person of population) {
        const key = attributeFunction(person);

        if (!strata.has(key)) {
            strata.set(key, []);
        }

        strata.get(key).push(person);
    }

    const allocations = new Map();
    let allocated = 0;

    for (const [key, units] of strata.entries()) {
        const allocation = Math.floor(
            units.length / population.length * sampleSize
        );

        allocations.set(key, allocation);
        allocated += allocation;
    }

    // Distribute rounding remainder.
    const keys = [...strata.keys()];
    let cursor = 0;

    while (allocated < sampleSize) {
        const key = keys[cursor % keys.length];

        if (allocations.get(key) < strata.get(key).length) {
            allocations.set(key, allocations.get(key) + 1);
            allocated++;
        }

        cursor++;
    }

    const result = [];

    for (const [key, units] of strata.entries()) {
        result.push(
            ...random.sample(units, allocations.get(key))
        );
    }

    return random.shuffle(result);
}

// ---------------------------------------------------------------------------
// 7. CLUSTER SAMPLING
// ---------------------------------------------------------------------------

function groupBy(population, keyFunction) {
    const groups = new Map();

    for (const item of population) {
        const key = keyFunction(item);

        if (!groups.has(key)) {
            groups.set(key, []);
        }

        groups.get(key).push(item);
    }

    return groups;
}

function clusterSample(population, clusterCount, seed = 42) {
    const clusters = groupBy(population, person => person.cluster);

    if (clusterCount <= 0 || clusterCount > clusters.size) {
        throw new RangeError("Invalid cluster count.");
    }

    const random = new SeededRandom(seed);
    const names = random.sample([...clusters.keys()], clusterCount);
    const result = [];

    for (const name of names) {
        result.push(...clusters.get(name));
    }

    return result;
}

// ---------------------------------------------------------------------------
// 8. MULTISTAGE SAMPLING
// ---------------------------------------------------------------------------

function multistageSample(
    population,
    clusterCount,
    peoplePerCluster,
    seed = 42
) {
    const clusters = groupBy(population, person => person.cluster);

    if (clusterCount <= 0 || clusterCount > clusters.size) {
        throw new RangeError("Invalid cluster count.");
    }

    if (peoplePerCluster <= 0) {
        throw new RangeError("People per cluster must be positive.");
    }

    const random = new SeededRandom(seed);
    const selectedNames = random.sample(
        [...clusters.keys()],
        clusterCount
    );

    const result = [];

    for (const name of selectedNames) {
        const units = clusters.get(name);
        const count = Math.min(peoplePerCluster, units.length);

        result.push(...random.sample(units, count));
    }

    return result;
}

// ---------------------------------------------------------------------------
// 9. NON-PROBABILITY SAMPLING
// ---------------------------------------------------------------------------

function convenienceSample(population, sampleSize) {
    validateSampleSize(population.length, sampleSize);

    // Accessibility is the selection mechanism rather than randomization.
    return population.slice(0, sampleSize);
}

function quotaSample(population, sampleSize, attributeFunction, quotas) {
    const quotaTotal = Object.values(quotas)
        .reduce((sum, value) => sum + value, 0);

    if (quotaTotal !== sampleSize) {
        throw new RangeError("Quota total must equal sample size.");
    }

    const result = [];

    for (const [category, quota] of Object.entries(quotas)) {
        const candidates = population.filter(
            person => attributeFunction(person) === category
        );

        if (candidates.length < quota) {
            throw new RangeError(
                `Not enough units for quota ${category}.`
            );
        }

        result.push(...candidates.slice(0, quota));
    }

    return result;
}

// ---------------------------------------------------------------------------
// 10. ESTIMATION
// ---------------------------------------------------------------------------

function meanIncome(sample) {
    return arithmeticMean(sample.map(person => person.income));
}

function samplingError(populationMean, estimate) {
    return estimate - populationMean;
}

function weightedMean(observations) {
    if (observations.length === 0) {
        throw new Error("Observations cannot be empty.");
    }

    const totalWeight = observations
        .reduce((sum, observation) => sum + observation.weight, 0);

    if (totalWeight <= 0) {
        throw new Error("Total weight must be positive.");
    }

    return observations.reduce(
        (sum, observation) =>
            sum + observation.value * observation.weight,
        0
    ) / totalWeight;
}

// ---------------------------------------------------------------------------
// 11. SAMPLE SIZE
// ---------------------------------------------------------------------------

function cochranSampleSize(
    z,
    proportion,
    marginOfError,
    populationSize = null
) {
    if (z <= 0) {
        throw new RangeError("Z value must be positive.");
    }

    if (proportion <= 0 || proportion >= 1) {
        throw new RangeError("Proportion must be between 0 and 1.");
    }

    if (marginOfError <= 0 || marginOfError >= 1) {
        throw new RangeError(
            "Margin of error must be between 0 and 1."
        );
    }

    const n0 =
        (z ** 2 * proportion * (1 - proportion))
        / marginOfError ** 2;

    if (populationSize === null) {
        return Math.ceil(n0);
    }

    if (populationSize <= 0) {
        throw new RangeError("Population size must be positive.");
    }

    const corrected =
        n0 / (1 + (n0 - 1) / populationSize);

    return Math.min(populationSize, Math.ceil(corrected));
}

// ---------------------------------------------------------------------------
// 12. CONFIDENCE INTERVAL
// ---------------------------------------------------------------------------

function confidenceIntervalForMean(values, z = 1.96) {
    if (values.length < 2) {
        throw new Error(
            "At least two observations are required."
        );
    }

    const average = arithmeticMean(values);
    const standardError =
        sampleStandardDeviation(values) / Math.sqrt(values.length);

    const margin = z * standardError;

    return {
        mean: average,
        standardError,
        lower: average - margin,
        upper: average + margin
    };
}

// ---------------------------------------------------------------------------
// 13. SAMPLING DISTRIBUTION
// ---------------------------------------------------------------------------

function simulateSamplingDistribution(
    population,
    sampleSize,
    repetitions = 200
) {
    if (repetitions <= 0) {
        throw new RangeError("Repetitions must be positive.");
    }

    const estimates = [];

    for (let i = 0; i < repetitions; i++) {
        const sample = simpleRandomSample(
            population,
            sampleSize,
            100 + i
        );

        estimates.push(meanIncome(sample));
    }

    return {
        meanOfEstimates: arithmeticMean(estimates),
        standardDeviationOfEstimates:
            sampleStandardDeviation(estimates),
        minimum: Math.min(...estimates),
        maximum: Math.max(...estimates)
    };
}

// ---------------------------------------------------------------------------
// 14. REPORTING
// ---------------------------------------------------------------------------

function printSection(title) {
    console.log(`\n${"=".repeat(60)}`);
    console.log(title);
    console.log("=".repeat(60));
}

function run() {
    const population = buildPopulation(3000, 42);
    const populationMean = arithmeticMean(
        population.map(person => person.income)
    );

    printSection("POPULATION");
    console.log(`Population size: ${population.length}`);
    console.log(
        `Population mean income: ${populationMean.toFixed(2)}`
    );

    printSection("SIMPLE RANDOM SAMPLING");
    const randomSample = simpleRandomSample(population, 200, 10);
    const randomEstimate = meanIncome(randomSample);

    console.log(`Sample size: ${randomSample.length}`);
    console.log(`Estimate: ${randomEstimate.toFixed(2)}`);
    console.log(
        `Sampling error: ${
            samplingError(populationMean, randomEstimate).toFixed(2)
        }`
    );

    printSection("SYSTEMATIC SAMPLING");
    const systematic = systematicSample(population, 200, 20);
    console.log(
        `Estimate: ${meanIncome(systematic).toFixed(2)}`
    );

    printSection("STRATIFIED SAMPLING");
    const stratified = stratifiedSample(
        population,
        200,
        person => person.region,
        30
    );

    console.log(
        `Estimate: ${meanIncome(stratified).toFixed(2)}`
    );

    const regionalComposition = groupBy(
        stratified,
        person => person.region
    );

    for (const [region, units] of regionalComposition) {
        console.log(`${region}: ${units.length}`);
    }

    printSection("CLUSTER SAMPLING");
    const clusters = clusterSample(population, 3, 40);
    console.log(`Observed units: ${clusters.length}`);
    console.log(
        `Estimate: ${meanIncome(clusters).toFixed(2)}`
    );

    printSection("MULTISTAGE SAMPLING");
    const multistage = multistageSample(
        population,
        5,
        20,
        50
    );

    console.log(`Final sample size: ${multistage.length}`);
    console.log(
        `Estimate: ${meanIncome(multistage).toFixed(2)}`
    );

    printSection("NON-PROBABILITY SAMPLING");
    const convenience = convenienceSample(
        population,
        100
    );

    console.log(
        `Convenience estimate: ${meanIncome(convenience).toFixed(2)}`
    );

    const quota = quotaSample(
        population,
        100,
        person => person.region,
        {
            North: 35,
            South: 25,
            East: 20,
            West: 20
        }
    );

    console.log(
        `Quota estimate: ${meanIncome(quota).toFixed(2)}`
    );

    printSection("SAMPLE SIZE");
    console.log(
        `Required sample: ${cochranSampleSize(
            1.96,
            0.5,
            0.05,
            population.length
        )}`
    );

    printSection("CONFIDENCE INTERVAL");
    const interval = confidenceIntervalForMean(
        randomSample.map(person => person.income)
    );

    console.log(`Mean: ${interval.mean.toFixed(2)}`);
    console.log(`SE: ${interval.standardError.toFixed(2)}`);
    console.log(
        `Approximate 95% CI: ${
            interval.lower.toFixed(2)
        } to ${
            interval.upper.toFixed(2)
        }`
    );

    printSection("WEIGHTED ESTIMATION");
    const weighted = weightedMean([
        { value: 100, weight: 1 },
        { value: 200, weight: 2 },
        { value: 300, weight: 1 }
    ]);

    console.log(`Weighted mean: ${weighted.toFixed(2)}`);

    printSection("SAMPLING DISTRIBUTION");
    const distribution = simulateSamplingDistribution(
        population,
        100,
        200
    );

    console.log(
        `Mean of sample means: ${
            distribution.meanOfEstimates.toFixed(2)
        }`
    );
    console.log(
        `SD of sample means: ${
            distribution.standardDeviationOfEstimates.toFixed(2)
        }`
    );

    printSection("EDGE CASE");
    try {
        simpleRandomSample(population, population.length + 1);
    } catch (error) {
        console.log(`Validation caught: ${error.message}`);
    }

    printSection("INTERPRETATION");
    console.log(
        "A sample statistic estimates a population parameter."
    );
    console.log(
        "Randomization controls selection probability; it does not eliminate every source of error."
    );
    console.log(
        "A representative-looking sample is not automatically a probability sample."
    );
}

run();

File: day_16_sampling/day_16_sampling.cpp
/*
 * Sampling: Population, Samples and Sampling Techniques
 *
 * C++17 case study:
 * A municipal research department wants to estimate average monthly
 * household income across a large city without surveying every household.
 *
 * The program models:
 * - A finite population
 * - Geographic clusters
 * - Stratification by region
 * - Simple random sampling
 * - Multistage sampling
 * - Estimation
 * - Sampling error
 * - Confidence intervals
 * - Sample-size planning
 * - Validation and failure conditions
 *
 * Compile:
 *   g++ -std=c++17 -O2 sampling_case_study.cpp -o sampling_case_study
 */

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <map>
#include <numeric>
#include <random>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

struct Household {
    int id;
    int age;
    std::string region;
    std::string employment;
    double monthlyIncome;
    std::string cluster;
};

using Population = std::vector<Household>;

class SamplingEngine {
private:
    std::mt19937 generator;

public:
    explicit SamplingEngine(unsigned seed = 42)
        : generator(seed) {}

    // Generate an integer from an inclusive range.
    int randomInteger(int minimum, int maximum) {
        std::uniform_int_distribution<int> distribution(
            minimum,
            maximum
        );

        return distribution(generator);
    }

    // Select k distinct elements without replacement.
    template <typename T>
    std::vector<T> randomSample(
        const std::vector<T>& source,
        std::size_t k
    ) {
        if (k > source.size()) {
            throw std::invalid_argument(
                "Sample size exceeds source size."
            );
        }

        std::vector<T> copy = source;

        std::shuffle(
            copy.begin(),
            copy.end(),
            generator
        );

        copy.resize(k);
        return copy;
    }
};

// ---------------------------------------------------------------------------
// Basic statistics
// ---------------------------------------------------------------------------

double mean(const std::vector<double>& values) {
    if (values.empty()) {
        throw std::invalid_argument(
            "Cannot calculate mean of empty data."
        );
    }

    return std::accumulate(
        values.begin(),
        values.end(),
        0.0
    ) / static_cast<double>(values.size());
}

double sampleStandardDeviation(
    const std::vector<double>& values
) {
    if (values.size() < 2) {
        return 0.0;
    }

    const double average = mean(values);

    double squaredDeviation = 0.0;

    for (double value : values) {
        squaredDeviation +=
            std::pow(value - average, 2);
    }

    return std::sqrt(
        squaredDeviation
        / static_cast<double>(values.size() - 1)
    );
}

std::vector<double> incomes(
    const std::vector<Household>& households
) {
    std::vector<double> values;
    values.reserve(households.size());

    for (const auto& household : households) {
        values.push_back(household.monthlyIncome);
    }

    return values;
}

// ---------------------------------------------------------------------------
// Synthetic city population
// ---------------------------------------------------------------------------

Population createPopulation(
    std::size_t size,
    unsigned seed
) {
    if (size == 0) {
        throw std::invalid_argument(
            "Population cannot be empty."
        );
    }

    SamplingEngine random(seed);

    const std::vector<std::string> regions = {
        "North",
        "South",
        "East",
        "West"
    };

    const std::vector<std::string> employmentTypes = {
        "Student",
        "Employed",
        "Self-employed",
        "Unemployed"
    };

    Population population;
    population.reserve(size);

    std::mt19937 normalGenerator(seed + 1);

    for (std::size_t i = 0; i < size; ++i) {
        const int age =
            random.randomInteger(18, 75);

        const std::string region =
            regions[i % regions.size()];

        const std::string employment =
            employmentTypes[
                random.randomInteger(
                    0,
                    static_cast<int>(employmentTypes.size()) - 1
                )
            ];

        double baseIncome = 0.0;

        if (employment == "Student") {
            baseIncome = 12000.0;
        } else if (employment == "Employed") {
            baseIncome = 50000.0;
        } else if (employment == "Self-employed") {
            baseIncome = 65000.0;
        } else {
            baseIncome = 10000.0;
        }

        std::normal_distribution<double> incomeDistribution(
            baseIncome,
            baseIncome * 0.25
        );

        double income =
            std::max(
                0.0,
                incomeDistribution(normalGenerator)
            );

        const int clusterNumber =
            static_cast<int>(i % 20) + 1;

        population.push_back(
            Household{
                static_cast<int>(i + 1),
                age,
                region,
                employment,
                income,
                "Ward-" +
                    std::to_string(clusterNumber)
            }
        );
    }

    return population;
}

// ---------------------------------------------------------------------------
// Simple random sampling
// ---------------------------------------------------------------------------

Population simpleRandomSample(
    const Population& population,
    std::size_t sampleSize,
    unsigned seed
) {
    if (sampleSize == 0 ||
        sampleSize > population.size()) {
        throw std::invalid_argument(
            "Invalid simple random sample size."
        );
    }

    SamplingEngine engine(seed);

    return engine.randomSample(
        population,
        sampleSize
    );
}

// ---------------------------------------------------------------------------
// Stratified random sampling
// ---------------------------------------------------------------------------

Population stratifiedSample(
    const Population& population,
    std::size_t sampleSize,
    unsigned seed
) {
    if (sampleSize == 0 ||
        sampleSize > population.size()) {
        throw std::invalid_argument(
            "Invalid stratified sample size."
        );
    }

    std::map<std::string, Population> strata;

    for (const auto& household : population) {
        strata[household.region].push_back(
            household
        );
    }

    SamplingEngine engine(seed);

    Population result;

    /*
     * Proportionate allocation:
     *
     * n_h = (N_h / N) * n
     *
     * The remaining observations after integer rounding are assigned
     * to the largest strata that still have capacity.
     */
    std::map<std::string, std::size_t> allocation;

    std::size_t allocated = 0;

    for (const auto& [region, units] : strata) {
        const double exactAllocation =
            static_cast<double>(units.size())
            / static_cast<double>(population.size())
            * static_cast<double>(sampleSize);

        const std::size_t count =
            static_cast<std::size_t>(
                std::floor(exactAllocation)
            );

        allocation[region] = count;
        allocated += count;
    }

    while (allocated < sampleSize) {
        auto candidate = std::max_element(
            allocation.begin(),
            allocation.end(),
            [&](const auto& left, const auto& right) {
                const double leftRate =
                    static_cast<double>(
                        strata.at(left.first).size()
                    ) / population.size();

                const double rightRate =
                    static_cast<double>(
                        strata.at(right.first).size()
                    ) / population.size();

                return leftRate < rightRate;
            }
        );

        if (candidate == allocation.end()) {
            break;
        }

        if (candidate->second <
            strata.at(candidate->first).size()) {
            ++candidate->second;
            ++allocated;
        } else {
            break;
        }
    }

    for (const auto& [region, units] : strata) {
        const std::size_t count =
            allocation[region];

        const Population selected =
            engine.randomSample(units, count);

        result.insert(
            result.end(),
            selected.begin(),
            selected.end()
        );
    }

    std::shuffle(
        result.begin(),
        result.end(),
        std::mt19937(seed + 100)
    );

    return result;
}

// ---------------------------------------------------------------------------
// Two-stage cluster sampling
// ---------------------------------------------------------------------------

Population multistageSample(
    const Population& population,
    std::size_t numberOfClusters,
    std::size_t householdsPerCluster,
    unsigned seed
) {
    if (numberOfClusters == 0 ||
        householdsPerCluster == 0) {
        throw std::invalid_argument(
            "Stage sizes must be positive."
        );
    }

    std::map<std::string, Population> clusters;

    for (const auto& household : population) {
        clusters[household.cluster].push_back(
            household
        );
    }

    if (numberOfClusters > clusters.size()) {
        throw std::invalid_argument(
            "Requested more clusters than exist."
        );
    }

    SamplingEngine engine(seed);

    std::vector<std::string> clusterNames;

    for (const auto& [name, units] : clusters) {
        clusterNames.push_back(name);
    }

    clusterNames =
        engine.randomSample(
            clusterNames,
            numberOfClusters
        );

    Population result;

    for (const auto& clusterName : clusterNames) {
        const auto& units =
            clusters.at(clusterName);

        const std::size_t count =
            std::min(
                householdsPerCluster,
                units.size()
            );

        const Population selected =
            engine.randomSample(units, count);

        result.insert(
            result.end(),
            selected.begin(),
            selected.end()
        );
    }

    return result;
}

// ---------------------------------------------------------------------------
// Estimation
// ---------------------------------------------------------------------------

double estimateMeanIncome(
    const Population& sample
) {
    return mean(incomes(sample));
}

double samplingError(
    double populationMean,
    double estimate
) {
    return estimate - populationMean;
}

double standardErrorOfMean(
    const Population& sample
) {
    const std::vector<double> values =
        incomes(sample);

    return sampleStandardDeviation(values)
        / std::sqrt(
            static_cast<double>(values.size())
        );
}

struct ConfidenceInterval {
    double estimate;
    double standardError;
    double lower;
    double upper;
};

ConfidenceInterval confidenceInterval(
    const Population& sample,
    double zValue = 1.96
) {
    if (sample.size() < 2) {
        throw std::invalid_argument(
            "At least two observations are required."
        );
    }

    const double estimate =
        estimateMeanIncome(sample);

    const double standardError =
        standardErrorOfMean(sample);

    const double margin =
        zValue * standardError;

    return {
        estimate,
        standardError,
        estimate - margin,
        estimate + margin
    };
}

// ---------------------------------------------------------------------------
// Cochran sample-size calculation
// ---------------------------------------------------------------------------

std::size_t cochranSampleSize(
    double z,
    double proportion,
    double marginOfError,
    std::size_t populationSize
) {
    if (z <= 0 ||
        proportion <= 0 ||
        proportion >= 1 ||
        marginOfError <= 0 ||
        marginOfError >= 1 ||
        populationSize == 0) {
        throw std::invalid_argument(
            "Invalid sample-size parameters."
        );
    }

    const double n0 =
        (z * z * proportion * (1.0 - proportion))
        / (marginOfError * marginOfError);

    const double corrected =
        n0
        / (
            1.0
            + (n0 - 1.0)
            / static_cast<double>(populationSize)
        );

    return std::min(
        populationSize,
        static_cast<std::size_t>(
            std::ceil(corrected)
        )
    );
}

// ---------------------------------------------------------------------------
// Repeated-sampling experiment
// ---------------------------------------------------------------------------

std::vector<double> repeatedSampling(
    const Population& population,
    std::size_t sampleSize,
    std::size_t repetitions
) {
    if (repetitions == 0) {
        throw std::invalid_argument(
            "Repetitions must be positive."
        );
    }

    std::vector<double> estimates;
    estimates.reserve(repetitions);

    for (std::size_t i = 0; i < repetitions; ++i) {
        const Population sample =
            simpleRandomSample(
                population,
                sampleSize,
                static_cast<unsigned>(1000 + i)
            );

        estimates.push_back(
            estimateMeanIncome(sample)
        );
    }

    return estimates;
}

// ---------------------------------------------------------------------------
// Utility display
// ---------------------------------------------------------------------------

void printResult(
    const std::string& name,
    double populationMean,
    const Population& sample
) {
    const double estimate =
        estimateMeanIncome(sample);

    std::cout
        << std::left
        << std::setw(28)
        << name
        << "n = "
        << std::setw(5)
        << sample.size()
        << " estimate = "
        << std::fixed
        << std::setprecision(2)
        << estimate
        << " error = "
        << samplingError(
            populationMean,
            estimate
        )
        << '\n';
}

// ---------------------------------------------------------------------------
// Main industry-style case study
// ---------------------------------------------------------------------------

int main() {
    try {
        std::cout
            << "SAMPLING CASE STUDY\n"
            << "Municipal household income survey\n"
            << "========================================\n\n";

        // The full population is available to the simulation but would
        // represent a costly census in the real-world scenario.
        const Population population =
            createPopulation(10000, 42);

        const double truePopulationMean =
            mean(incomes(population));

        std::cout
            << "Population size: "
            << population.size()
            << '\n'
            << "True population mean income: "
            << std::fixed
            << std::setprecision(2)
            << truePopulationMean
            << "\n\n";

        // Strategy 1: simple random sample.
        const Population randomSample =
            simpleRandomSample(
                population,
                400,
                10
            );

        printResult(
            "Simple random sampling",
            truePopulationMean,
            randomSample
        );

        // Strategy 2: stratified sample.
        const Population stratified =
            stratifiedSample(
                population,
                400,
                20
            );

        printResult(
            "Stratified sampling",
            truePopulationMean,
            stratified
        );

        // Strategy 3: two-stage cluster sample.
        const Population multistage =
            multistageSample(
                population,
                8,
                50,
                30
            );

        printResult(
            "Two-stage cluster sampling",
            truePopulationMean,
            multistage
        );

        // Confidence interval for the stratified sample.
        const ConfidenceInterval interval =
            confidenceInterval(stratified);

        std::cout
            << "\nAPPROXIMATE 95% CONFIDENCE INTERVAL\n"
            << "Estimate: "
            << interval.estimate
            << '\n'
            << "Standard error: "
            << interval.standardError
            << '\n'
            << "Lower bound: "
            << interval.lower
            << '\n'
            << "Upper bound: "
            << interval.upper
            << "\n\n";

        // Sample-size planning.
        const std::size_t plannedSampleSize =
            cochranSampleSize(
                1.96,
                0.50,
                0.05,
                population.size()
            );

        std::cout
            << "PLANNED SAMPLE SIZE\n"
            << "For 95% confidence, p=0.50, "
               "and 5% margin of error: "
            << plannedSampleSize
            << "\n\n";

        // Repeated sampling shows the sampling distribution of the mean.
        const auto estimates =
            repeatedSampling(
                population,
                100,
                300
            );

        const double averageEstimate =
            mean(estimates);

        const double distributionSD =
            sampleStandardDeviation(estimates);

        std::cout
            << "SAMPLING DISTRIBUTION EXPERIMENT\n"
            << "Repeated samples: "
            << estimates.size()
            << '\n'
            << "Mean of sample means: "
            << averageEstimate
            << '\n'
            << "SD of sample means: "
            << distributionSD
            << '\n';

        /*
         * Edge-case validation:
         *
         * A production sampling system must reject invalid designs instead
         * of silently producing misleading results.
         */
        std::cout
            << "\nVALIDATION TEST\n";

        try {
            simpleRandomSample(
                population,
                population.size() + 1,
                99
            );
        } catch (const std::exception& error) {
            std::cout
                << "Invalid sample correctly rejected: "
                << error.what()
                << '\n';
        }

        /*
         * Design trade-off:
         *
         * Simple random sampling is conceptually clean but may be expensive
         * when sampled units are geographically dispersed.
         *
         * Stratification can improve representation and precision when
         * strata are appropriately defined.
         *
         * Cluster sampling can reduce travel and operational cost, but
         * observations inside clusters may be correlated. That correlation
         * can increase variance relative to a simple random sample of the
         * same nominal size.
         *
         * Multistage sampling combines operational efficiency with
         * probability-based selection, but its variance calculation requires
         * accounting for the complete design rather than treating the final
         * observations as a simple random sample.
         */

        std::cout
            << "\nDESIGN INTERPRETATION\n"
            << "Population -> sampling frame -> clusters/strata -> "
               "selected units -> measurements -> estimator\n";

        std::cout
            << "\nCASE STUDY COMPLETE\n";
    }
    catch (const std::exception& error) {
        std::cerr
            << "Fatal error: "
            << error.what()
            << '\n';

        return 1;
    }

    return 0;
}
